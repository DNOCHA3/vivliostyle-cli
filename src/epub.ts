import { XMLParser } from 'fast-xml-parser';
import fs from 'node:fs';
import path from 'upath';
import { inflateZip, useTmpDirectory } from './util.js';

const pickFirstOne = <T>(arg: T | T[]): T =>
  Array.isArray(arg) ? arg[0] : arg;

export async function openEpubToTmpDirectory(filePath: string): Promise<{
  dest: string;
  epubOpfPath: string;
  deleteEpub: () => void;
}> {
  const [tmpDir, deleteEpub] = await useTmpDirectory();
  await inflateZip(filePath, tmpDir);

  const containerXmlPath = path.join(tmpDir, 'META-INF/container.xml');
  const xmlParser = new XMLParser({
    ignoreAttributes: false,
  });
  const { container } = xmlParser.parse(
    fs.readFileSync(containerXmlPath, 'utf8'),
  );
  const rootfile = pickFirstOne(container.rootfiles.rootfile); // Only supports a default rendition
  const epubOpfPath = path.join(tmpDir, rootfile['@_full-path']);
  return { dest: tmpDir, epubOpfPath, deleteEpub };
}
