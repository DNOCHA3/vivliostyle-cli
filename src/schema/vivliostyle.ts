import { createRequire } from 'node:module';

// TODO: Change to static import after JSON module becomes stable
const require = createRequire(import.meta.url);

const vivliostyleConfigSchema = require('../../schemas/vivliostyle/vivliostyleConfig.schema.json');

export { vivliostyleConfigSchema };

export type StructuredDocument = {
  title: string;
  href: string;
  sections?: StructuredDocumentSection[];
  children: StructuredDocument[];
};

export type StructuredDocumentSection = {
  headingHtml: string;
  headingText: string;
  href?: string;
  level?: number;
  id?: string;
  children: StructuredDocumentSection[];
};
