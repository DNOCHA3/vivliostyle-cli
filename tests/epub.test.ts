import { fs as memfs, vol } from 'memfs';
import { format } from 'prettier';
import tmp from 'tmp';
import { afterEach, expect, it, vi } from 'vitest';
import { exportEpub } from '../src/output/epub.js';
import { PublicationManifest } from '../src/schema/publication.schema.js';
import { toTree } from './commandUtil.js';

vi.mock('node:fs', () => ({ ...memfs, default: memfs }));

vi.mock('@vivliostyle/jsdom', () =>
  import('./commandUtil.js').then(({ getMockedJSDOM }) => getMockedJSDOM()),
);

vi.mock('tmp', () => {
  const mod = {
    __count: 0,
    dir: (_, cb) => {
      const target = `/tmp/${++mod.__count}`;
      memfs.mkdirSync(target, { recursive: true });
      cb(null, target);
    },
  };
  return { default: mod };
});

afterEach(() => {
  vol.reset();
  (tmp as any).__count = 0;
});

it('generate EPUB from single HTML with pub manifest', async () => {
  const manifest: PublicationManifest = {
    '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
    conformsTo: 'yuno',
    readingProgression: 'rtl',
    resources: [
      { url: 'cover.png', rel: 'cover' },
      { url: 'index.html', rel: 'pagelist' },
    ],
    author: {
      name: [
        { value: 'Kenji Miyazawa', language: 'en' },
        { value: '宮沢賢治', language: 'ja' },
      ],
    },
    creator: ['foo', 'bar'],
    editor: 'baz',
    artist: 'a',
    illustrator: 'i',
    colorist: 'x',
    penciler: 'q',
    inker: 'c',
    letterer: 'b',
    translator: 'l',
    readBy: 'o',
    publisher: { name: 'Publisher', type: 'Organization' },
    contributor: 'Contributor',
    copyrightHolder: 'Acme Corporation',
    copyrightYear: '2023',
    subject: 'Subject',
  };
  vol.fromJSON({
    '/work/input/publication.json': JSON.stringify(manifest),
    '/work/input/index.html': /* html */ `
      <html lang="ja-JP">
      <head>
        <title>Document</title>
        <link rel="stylesheet" type="text/css" href="assets/style.css">
        <link rel="publication" type="application/ld+json" href="publication.json">
      </head>
      <body>
        <div role="doc-toc">
          <h1>Table of Contents</h1>
          <ol>
            <li><a href="#intro">Intro</a></li>
            <li>
              <a href="#main">Main</a>
              <ol>
                <li><a href="#main-1">Main 1</a></li>
                <li><a href="#main-2">Main 2</a></li>
              </ol>
            </li>
          </ol>
        </div>
        <div role="doc-pagelist">
          <ol>
            <li><a href="#intro">Intro</a></li>
            <li><a href="#main">Main</a></li>
            <li><a href="#main-1">Main 1</a></li>
            <li><a href="#main-2">Main 2</a></li>
          </ol>
        </div>
        <math></math>
        <svg></svg>
        <script src="https://example.com/remote-script.js"></script>
      </body>
      </html>
    `,
    '/work/input/cover.png': '',
  });
  await exportEpub({
    webpubDir: '/work/input',
    entryHtmlFile: 'index.html',
    entryContextUrl: 'file://work/input/publication.json',
    manifest,
    target: '/work/output.epub',
    epubVersion: '3.0',
  });

  expect(toTree(vol)).toMatchSnapshot('tree');
  const file = vol.toJSON();
  expect(file['/tmp/1/META-INF/container.xml']).toMatchSnapshot(
    'container.xml',
  );
  expect(
    file['/tmp/1/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  expect(
    file['/tmp/1/EPUB/toc.ncx']?.replace(
      /<meta name="dtb:uid" content=".+"><\/meta>/g,
      '',
    ),
  ).toMatchSnapshot('toc.ncx');
  const entry = file['/tmp/1/EPUB/index.xhtml'];
  expect(format(entry as string, { parser: 'html' })).toMatchSnapshot(
    'index.xhtml',
  );
});

it('generate EPUB from series of HTML files', async () => {
  const manifest: PublicationManifest = {
    '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
    conformsTo: 'yuno',
    readingOrder: ['src/index.html', 'src/a/index.html', 'src/b/c/d.html'],
  };
  vol.fromJSON({
    '/work/input/src/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>My book</title>
      </head>
      <body>
        <a href="#foo">1</a>
        <a href="a">2</a>
        <a href="./b/c/d/">3</a>
      </body>
      </html>
    `,
    '/work/input/src/a/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>yuno</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/src/b/c/d.html': /* html */ `
      <html lang="en">
      <head>
        <title>yunocchi</title>
      </head>
      <body></body>
      </html>
    `,
  });
  await exportEpub({
    webpubDir: '/work/input',
    manifest,
    target: '/work/output.epub',
    epubVersion: '3.0',
  });

  expect(toTree(vol)).toMatchSnapshot('tree');
  const file = vol.toJSON();
  expect(
    file['/tmp/1/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  expect(
    file['/tmp/1/EPUB/toc.ncx']?.replace(
      /<meta name="dtb:uid" content=".+"><\/meta>/g,
      '',
    ),
  ).toMatchSnapshot('toc.ncx');
  const first = file['/tmp/1/EPUB/src/index.xhtml'];
  expect(format(first as string, { parser: 'html' })).toMatchSnapshot(
    'src/index.xhtml',
  );
});

it('generate EPUB with a hidden cover document', async () => {
  const manifest: PublicationManifest = {
    '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
    conformsTo: 'yuno',
    readingOrder: ['src/index.html', 'src/a.html'],
    resources: [
      {
        rel: 'cover',
        type: 'LinkedResource',
        url: 'cover.html',
        encodingFormat: 'text/html',
      },
      {
        rel: 'cover',
        url: 'cover.png',
        encodingFormat: 'image/png',
      },
    ],
  };
  vol.fromJSON({
    '/work/input/src/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>yuno</title>
      </head>
      <body>
        <nav role="doc-toc">
          <ol><li><a href="a.html">A</a></li></ol>
        </nav>
      </body>
      </html>
    `,
    '/work/input/src/a.html': /* html */ `<html lang="en"><head><title></title></head><body></body></html>`,
    '/work/input/cover.html': /* html */ `<html lang="en"><head><title></title></head><body></body></html>`,
    '/work/input/cover.png': '',
  });
  await exportEpub({
    webpubDir: '/work/input',
    manifest,
    target: '/work/output.epub',
    epubVersion: '3.0',
  });

  expect(toTree(vol)).toMatchSnapshot('tree');
  const file = vol.toJSON();
  expect(
    file['/tmp/1/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
});
