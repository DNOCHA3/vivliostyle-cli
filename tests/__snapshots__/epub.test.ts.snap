// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate EPUB from series of HTML files > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"en\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>My book</dc:title>
    <dc:language>en</dc:language>
    
  </metadata>
  <manifest>
    <item id=\\"srcindexxhtml\\" href=\\"src/index.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"></item>
    <item id=\\"srcaindexxhtml\\" href=\\"src/a/index.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"srcbcdxhtml\\" href=\\"src/b/c/d.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"tocncx\\" href=\\"toc.ncx\\" media-type=\\"application/x-dtbncx+xml\\"></item>
  </manifest>
  <spine toc=\\"tocncx\\">
    <itemref idref=\\"srcindexxhtml\\"></itemref>
    <itemref idref=\\"srcaindexxhtml\\"></itemref>
    <itemref idref=\\"srcbcdxhtml\\"></itemref>
  </spine>
  <guide>
    <reference type=\\"toc\\" href=\\"src/index.xhtml#toc\\" title=\\"Table of Contents\\"></reference>
  </guide>
</package>
"
`;

exports[`generate EPUB from series of HTML files > src/index.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  lang=\\"en\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
>
  <head>
    <title>My book</title>
  </head>
  <body>
    <a href=\\"index.xhtml#foo\\">1</a>
    <a href=\\"a\\">2</a>
    <a href=\\"b/c/d\\">3</a>

    <nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\">
      <ol>
        <li><a epub:type=\\"toc\\" href=\\"index.xhtml#toc\\">Table of Contents</a></li>
      </ol>
    </nav>
    <nav id=\\"toc\\" role=\\"doc-toc\\" epub:type=\\"toc\\" hidden=\\"\\">
      <ol>
        <li><a href=\\"index.xhtml\\">My book</a></li>
        <li><a href=\\"a/index.xhtml\\">yuno</a></li>
        <li><a href=\\"b/c/d.xhtml\\">yunocchi</a></li>
      </ol>
    </nav>
  </body>
</html>
"
`;

exports[`generate EPUB from series of HTML files > toc.ncx 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<ncx xmlns=\\"http://www.daisy.org/z3986/2005/ncx/\\" version=\\"2005-1\\">
  <head>
    
    <meta name=\\"dtb:depth\\" content=\\"1\\"></meta>
    <meta name=\\"dtb:totalPageCount\\" content=\\"0\\"></meta>
    <meta name=\\"dtb:maxPageNumber\\" content=\\"0\\"></meta>
  </head>
  <docTitle>
    <text>My book</text>
  </docTitle>
  <navMap>
    <navPoint id=\\"navpoint-1\\">
      <navLabel>
        <text>My book</text>
      </navLabel>
      <content src=\\"src/index.xhtml\\"></content>
    </navPoint>
    <navPoint id=\\"navpoint-2\\">
      <navLabel>
        <text>yuno</text>
      </navLabel>
      <content src=\\"src/a/index.xhtml\\"></content>
    </navPoint>
    <navPoint id=\\"navpoint-3\\">
      <navLabel>
        <text>yunocchi</text>
      </navLabel>
      <content src=\\"src/b/c/d.xhtml\\"></content>
    </navPoint>
  </navMap>
</ncx>
"
`;

exports[`generate EPUB from series of HTML files > tree 1`] = `
"/
├─ tmp/
│  └─ 1/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  ├─ src/
│     │  │  ├─ a/
│     │  │  │  └─ index.xhtml
│     │  │  ├─ b/
│     │  │  │  └─ c/
│     │  │  │     └─ d.xhtml
│     │  │  └─ index.xhtml
│     │  └─ toc.ncx
│     └─ META-INF/
│        └─ container.xml
└─ work/
   ├─ input/
   │  ├─ publication.json
   │  └─ src/
   │     ├─ a/
   │     │  └─ index.html
   │     ├─ b/
   │     │  └─ c/
   │     │     └─ d.html
   │     └─ index.html
   └─ output.epub"
`;

exports[`generate EPUB from single HTML with pub manifest > container.xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<container version=\\"1.0\\" xmlns=\\"urn:oasis:names:tc:opendocument:xmlns:container\\">
  <rootfiles>
    <rootfile full-path=\\"EPUB/content.opf\\" media-type=\\"application/oebps-package+xml\\" />
  </rootfiles>
</container>"
`;

exports[`generate EPUB from single HTML with pub manifest > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"ja-JP\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>Document</dc:title>
    <dc:language>ja-JP</dc:language>
    <dc:creator id=\\"author-1\\">宮沢賢治</dc:creator>
    <dc:creator id=\\"creator-1\\">foo</dc:creator>
    <dc:creator id=\\"creator-2\\">bar</dc:creator>
    <dc:creator id=\\"editor-1\\">baz</dc:creator>
    <dc:creator id=\\"artist-1\\">a</dc:creator>
    <dc:creator id=\\"illustrator-1\\">i</dc:creator>
    <dc:creator id=\\"colorist-1\\">x</dc:creator>
    <dc:creator id=\\"penciler-1\\">q</dc:creator>
    <dc:creator id=\\"inker-1\\">c</dc:creator>
    <dc:creator id=\\"letterer-1\\">b</dc:creator>
    <dc:creator id=\\"translator-1\\">l</dc:creator>
    <dc:creator id=\\"readby-1\\">o</dc:creator>
    <dc:publisher id=\\"publisher-1\\">Publisher</dc:publisher>
    <dc:contributor id=\\"contributor-1\\">Contributor</dc:contributor>
    <dc:rights>© 2023 Acme Corporation</dc:rights>
    <dc:subject>Subject</dc:subject>
    
    <meta name=\\"cover\\" content=\\"coverpng\\"></meta>
  </metadata>
  <manifest>
    <item id=\\"coverpng\\" href=\\"cover.png\\" media-type=\\"image/png\\" properties=\\"cover-image\\"></item>
    <item id=\\"indexxhtml\\" href=\\"index.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav mathml remote-resources scripted svg\\"></item>
    <item id=\\"tocncx\\" href=\\"toc.ncx\\" media-type=\\"application/x-dtbncx+xml\\"></item>
  </manifest>
  <spine toc=\\"tocncx\\" page-progression-direction=\\"rtl\\">
    <itemref idref=\\"indexxhtml\\"></itemref>
  </spine>
  <guide>
    <reference type=\\"toc\\" href=\\"index.xhtml#toc\\" title=\\"Table of Contents\\"></reference>
  </guide>
</package>
"
`;

exports[`generate EPUB from single HTML with pub manifest > index.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  lang=\\"ja-JP\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
>
  <head>
    <title>Document</title>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"assets/style.css\\" />
  </head>
  <body>
    <nav role=\\"doc-toc\\" id=\\"toc\\" epub:type=\\"toc\\">
      <h1>Table of Contents</h1>
      <ol>
        <li><a href=\\"index.xhtml#intro\\">Intro</a></li>
        <li>
          <a href=\\"index.xhtml#main\\">Main</a>
          <ol>
            <li><a href=\\"index.xhtml#main-1\\">Main 1</a></li>
            <li><a href=\\"index.xhtml#main-2\\">Main 2</a></li>
          </ol>
        </li>
      </ol>
    </nav>
    <nav role=\\"doc-pagelist\\" id=\\"page-list\\" epub:type=\\"page-list\\">
      <ol>
        <li><a href=\\"index.xhtml#intro\\">Intro</a></li>
        <li><a href=\\"index.xhtml#main\\">Main</a></li>
        <li><a href=\\"index.xhtml#main-1\\">Main 1</a></li>
        <li><a href=\\"index.xhtml#main-2\\">Main 2</a></li>
      </ol>
    </nav>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" />
    <svg xmlns=\\"http://www.w3.org/2000/svg\\" />
    <script src=\\"https://example.com/remote-script.js\\"></script>

    <nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\">
      <ol>
        <li><a epub:type=\\"toc\\" href=\\"index.xhtml#toc\\">Table of Contents</a></li>
      </ol>
    </nav>
  </body>
</html>
"
`;

exports[`generate EPUB from single HTML with pub manifest > toc.ncx 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<ncx xmlns=\\"http://www.daisy.org/z3986/2005/ncx/\\" version=\\"2005-1\\">
  <head>
    
    <meta name=\\"dtb:depth\\" content=\\"1\\"></meta>
    <meta name=\\"dtb:totalPageCount\\" content=\\"0\\"></meta>
    <meta name=\\"dtb:maxPageNumber\\" content=\\"0\\"></meta>
  </head>
  <docTitle>
    <text>Document</text>
  </docTitle>
  <navMap>
    <navPoint id=\\"navpoint-1\\">
      <navLabel>
        <text>Intro</text>
      </navLabel>
      <content src=\\"index.xhtml#intro\\"></content>
    </navPoint>
    <navPoint id=\\"navpoint-2\\">
      <navLabel>
        <text>Main</text>
      </navLabel>
      <content src=\\"index.xhtml#main\\"></content>
      <navPoint id=\\"navpoint-3\\">
        <navLabel>
          <text>Main 1</text>
        </navLabel>
        <content src=\\"index.xhtml#main-1\\"></content>
      </navPoint>
      <navPoint id=\\"navpoint-4\\">
        <navLabel>
          <text>Main 2</text>
        </navLabel>
        <content src=\\"index.xhtml#main-2\\"></content>
      </navPoint>
    </navPoint>
  </navMap>
</ncx>
"
`;

exports[`generate EPUB from single HTML with pub manifest > tree 1`] = `
"/
├─ tmp/
│  └─ 1/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  ├─ cover.png
│     │  ├─ index.xhtml
│     │  └─ toc.ncx
│     └─ META-INF/
│        └─ container.xml
└─ work/
   ├─ input/
   │  ├─ cover.png
   │  ├─ index.html
   │  └─ publication.json
   └─ output.epub"
`;
